packages/dashboard/src/components/delivery-alert-store.tsx

```diff
@@ -0,0 +1,539 @@
+          }
+          continue;
+        }
+
+        // No task involved, and still waiting for user action. There should not
+        // be any longstanding delivery alerts that appear after a refresh, only
+        // the delivery alerts that are currently present and have not been
+        // responded to.
+        if (!alert.task_id) {
+          filteredAlertsMap[alert.id] = { deliveryAlert: alert, taskState: undefined };
+          continue;
+        }
+
+        // Tasks that have not been encountered before will go into
+        // taskIdToAlertsMap first, as an error will supercede a warning.
+        if (!Object.keys(taskIdToAlertsMap).includes(alert.task_id)) {
+          let state: TaskState | undefined = undefined;
+          try {
+            state = (await rmf.tasksApi.getTaskStateTasksTaskIdStateGet(alert.task_id)).data;
+          } catch {
+            console.error(
+              `Failed to fetch task state for ${alert.task_id} for delivery alert ${alert.id}`,
+            );
+          }
+          taskIdToAlertsMap[alert.task_id] = {
+            deliveryAlert: alert,
+            taskState: state,
+          };
+        } else if (alert.tier === 'error') {
+          let state: TaskState | undefined = undefined;
+          try {
+            state = (await rmf.tasksApi.getTaskStateTasksTaskIdStateGet(alert.task_id)).data;
+          } catch {
+            console.error(
+              `Failed to fetch task state for ${alert.task_id} for delivery alert ${alert.id}`,
+            );
+          }
+
+          taskIdToAlertsMap[alert.task_id] = {
+            deliveryAlert: alert,
+            taskState: state,
+          };
+        }
+      }
+
+      // Move all unresolved and up-to-date task related delivery alerts to the
+      // filtered map.
+      for (const taskId of Object.keys(taskIdToAlertsMap)) {
+        if (resolvedTaskIds.includes(taskId)) {
+          continue;
+        }
+        const alertData = taskIdToAlertsMap[taskId];
+        filteredAlertsMap[alertData.deliveryAlert.id] = alertData;
+      }
+      setAlerts(filteredAlertsMap);
+    })();
+
+    const sub = rmf.deliveryAlertObsStore.subscribe(async (deliveryAlert) => {
+      let state: TaskState | undefined = undefined;
+      if (deliveryAlert.task_id) {
+        try {
+          state = (await rmf.tasksApi.getTaskStateTasksTaskIdStateGet(deliveryAlert.task_id)).data;
+        } catch {
+          console.error(`Failed to fetch task state for ${deliveryAlert.task_id}`);
+        }
+      }
+      filterAndPushDeliveryAlert(deliveryAlert, state);
+    });
+    return () => sub.unsubscribe();
+  }, [rmf]);
+
+  const onCancel = React.useCallback<Required<DeliveryWarningDialogProps>['onCancel']>(
+    async (delivery_alert_id, task_id) => {
+      try {
+        if (!rmf) {
+          throw new Error('tasks and delivery alert api not available');
+        }
+        await rmf.tasksApi?.postCancelTaskTasksCancelTaskPost({
```

Comment:
Maybe the '?' is not necessary

Reply:
https://github.com/open-rmf/rmf-web/pull/768/commits/83666bba825ec114cc679179300587c135db958a